syft/pkg/cataloger/generic/cataloger.go

```diff
@@ -23,6 +23,27 @@ type Cataloger struct {
 	upstreamCataloger string
 }
 
+func (c *Cataloger) WithParser(parser Parser, searchRequests ...SearchRequest) *Cataloger {
+	c.processor = append(c.processor,
+		func(resolver source.FileResolver, env Environment) []request {
+			var requests []request
+			for _, req := range searchRequests {
+				log.Tracef("searching for: %s", req.String())
+
+				matches, err := req.Execute(resolver)
+				if err != nil {
+					log.WithFields("error", err).Warnf("unable to process search request %q", req.String())
+					continue
+				}
+
+				requests = append(requests, makeRequests(parser, matches)...)
+			}
+			return requests
+		},
+	)
+	return c
+}
+
 func (c *Cataloger) WithParserByGlobs(parser Parser, globs ...string) *Cataloger {
```

Comment:
Last thought about the glob processing -- I think we'd just need to update this method to be smart about looking at the glob string and using the fastest indexes it can, falling back to glob matching if necessary. E.g. split the glob at the last `/` to get the basename, we'd have some obvious cases:
* basename has wildcards?
  * has suffix? use suffix matching to filter first
  * no suffix? use basename by glob
* no wildcards? use basename directly
* prefix = `**`? return
* post-filter by glob pattern

I'm sure there are more cases than this, but just the general idea

Reply:
unfortunately there are a lot more cases, but I agree with the overall point that we should try to push the indexing down as an implementation concern and do our best to leverage the correct indexes based off of string parsing. The one thing I don't like about that approach is that we're making specific behavioral hints through strings instead of concrete types / interfaces. That being said I still think I like the suggestion to do string globs.